### ISSUE TITLE
No grad for G, B of input material texture

### ISSUE THREAD

**Issue by nkyriazis** [BachiLi/redner#73]
I'm having some issues optimizing over a material texture from observations, so I made this example to show off the problem. I'm texturing a triangle with a pattern and try to reconstruct it back through gradient descend on the image difference to the target.

I'm not sure what the exact problem is, but, it's interesting how I'm getting no gradient for the GB channels of the texture. Am I doing something wrong?

Running (after some fixes) the example in #64 I see that the GB channels have not grad there either.

```python
import sys
sys.path.append('/home/kyriazis/projects/redner')
sys.path.append('/home/kyriazis/projects/redner/build')
```

```python
import pyredner
import torch
import torch.optim as optim
import matplotlib.pyplot as plt
from matplotlib import cm
import numpy as np
```

```python
pyredner.set_use_gpu(True)
pyredner.render_pytorch.print_timing = False
```

```python
colors = cm.jet(np.linspace(0, 1, 10 * 10))[...,:3].reshape(10, 10, 3)
```

```python
material = pyredner.Material(diffuse_reflectance=pyredner.Texture(torch.tensor(colors, device='cuda').float()))
ground_truth = material.diffuse_reflectance.texels.clone()
```

```python
tri = pyredner.Shape(vertices=torch.tensor([[-1., 1., 0], [1, 1, 0], [0, -1, 0]]).cuda(),
                     indices=torch.tensor([[0, 1, 2]]).int().cuda(), material_id=0,
                     uvs=torch.tensor([[0., 0.], [1., 0.], [0.5, 1.]]).cuda(),
                     uv_indices=torch.tensor([[0, 1, 2]]).int().cuda())
```

```python
envmap = pyredner.EnvironmentMap(0.7 * torch.ones(1, 1, 3).cuda())
```

```python
cam = pyredner.Camera(position=torch.tensor([0, 0, -2.]),
                      look_at=torch.zeros(3).float(),
                      up=torch.tensor([0., 1, 0]),
                      fov=torch.tensor([45.]))
```

```python
scene = pyredner.Scene(cam, [tri], [material], [], envmap)
```

```python
serialized = pyredner.RenderFunction.serialize_scene(scene, 16, 1,
                                                    use_primary_edge_sampling=False, use_secondary_edge_sampling=False)
```

```python
img = pyredner.RenderFunction.apply(0, *serialized)
```

```python
plt.figure(figsize=(10, 10))
plt.imshow(img.cpu())
```

    <matplotlib.image.AxesImage at 0x7f9674151ac8>

![output_11_1](https://user-images.githubusercontent.com/13269427/69569527-8525a700-0fc6-11ea-8cff-16d1e7db67bb.png)

```python
def model(x):
    material.diffuse_reflectance = pyredner.Texture(x)
    serialized = pyredner.RenderFunction.serialize_scene(scene, 16, 1,
                                                    use_primary_edge_sampling=False, use_secondary_edge_sampling=False)
    return pyredner.RenderFunction.apply(0, *serialized)
```

```python
color = torch.rand(10, 10, 3).cuda().requires_grad_(True)
```

```python
optimizer = optim.Adam([color], lr=0.1)
```

```python
for iteration in range(100):
    optimizer.zero_grad()
    gen = model(color)
    loss = ((gen - img)**2).mean()
    loss.backward()
    optimizer.step()
    print('Loss =', loss.item())
```

    Loss = 0.05630433186888695
    Loss = 0.06002412736415863
    Loss = 0.0571032389998436
    Loss = 0.05493753403425217
    Loss = 0.05425659567117691
    Loss = 0.05458734557032585
    Loss = 0.05549685284495354
    Loss = 0.05669301748275757
    Loss = 0.057747673243284225
    Loss = 0.058230236172676086
    Loss = 0.05803096294403076
    Loss = 0.05735134333372116
    Loss = 0.05649701878428459
    Loss = 0.05574801564216614
    Loss = 0.05528859794139862
    Loss = 0.05518066883087158
    Loss = 0.055393874645233154
    Loss = 0.05584738776087761
    Loss = 0.05641963332891464
    Loss = 0.056950341910123825
    Loss = 0.05727851390838623
    Loss = 0.05731093883514404
    Loss = 0.057063184678554535
    Loss = 0.05664169788360596
    Loss = 0.05619248002767563
    Loss = 0.05584871023893356
    Loss = 0.05569377541542053
    Loss = 0.05574836581945419
    Loss = 0.055977992713451385
    Loss = 0.05630658566951752
    Loss = 0.056631963700056076
    Loss = 0.05685194954276085
    Loss = 0.05690034478902817
    Loss = 0.05677306652069092
    Loss = 0.056526925414800644
    Loss = 0.05625336617231369
    Loss = 0.056042950600385666
    Loss = 0.055956076830625534
    Loss = 0.05600859969854355
    Loss = 0.05617235228419304
    Loss = 0.05638617277145386
    Loss = 0.056574754416942596
    Loss = 0.056673839688301086
    Loss = 0.056654393672943115
    Loss = 0.05653296038508415
    Loss = 0.0563623309135437
    Loss = 0.05620831996202469
    Loss = 0.056124500930309296
    Loss = 0.056134697049856186
    Loss = 0.056227460503578186
    Loss = 0.056362323462963104
    Loss = 0.05648518726229668
    Loss = 0.05654922127723694
    Loss = 0.05653379485011101
    Loss = 0.05645202845335007
    Loss = 0.056342750787734985
    Loss = 0.0562518946826458
    Loss = 0.05621325969696045
    Loss = 0.05623685568571091
    Loss = 0.0563075989484787
    Loss = 0.05639313906431198
    Loss = 0.0564572773873806
    Loss = 0.056474827229976654
    Loss = 0.05644198879599571
    Loss = 0.056377094238996506
    Loss = 0.05631104111671448
    Loss = 0.05627257749438286
    Loss = 0.05627582594752312
    Loss = 0.056315552443265915
    Loss = 0.056371234357357025
    Loss = 0.056417085230350494
    Loss = 0.05643363296985626
    Loss = 0.056415803730487823
    Loss = 0.05637430399656296
    Loss = 0.05632978305220604
    Loss = 0.0563027486205101
    Loss = 0.0563041977584362
    Loss = 0.05633116886019707
    Loss = 0.0563691183924675
    Loss = 0.05639946088194847
    Loss = 0.05640862137079239
    Loss = 0.0563940666615963
    Loss = 0.05636468529701233
    Loss = 0.056335628032684326
    Loss = 0.05632065236568451
    Loss = 0.05632568895816803
    Loss = 0.05634660646319389
    Loss = 0.056371938437223434
    Loss = 0.056388966739177704
    Loss = 0.05639012157917023
    Loss = 0.056376270949840546
    Loss = 0.05635557696223259
    Loss = 0.056338872760534286
    Loss = 0.05633416771888733
    Loss = 0.05634289234876633
    Loss = 0.05635957047343254
    Loss = 0.05637499317526817
    Loss = 0.056381236761808395
    Loss = 0.056375764310359955
    Loss = 0.05636243522167206

```python
plt.imshow(model(color).cpu().detach())
```

    Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).

    <matplotlib.image.AxesImage at 0x7f96741a9f60>

![output_16_2](https://user-images.githubusercontent.com/13269427/69569528-85be3d80-0fc6-11ea-9c27-bc06acfe1729.png)

```python
plt.imshow(ground_truth.cpu())
```

    <matplotlib.image.AxesImage at 0x7f966c301390>

![output_17_1](https://user-images.githubusercontent.com/13269427/69569531-85be3d80-0fc6-11ea-8dab-e999ec14312d.png)

```python
plt.imshow(color.detach().cpu())
```

    Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).

    <matplotlib.image.AxesImage at 0x7f966c2cccf8>

![output_18_2](https://user-images.githubusercontent.com/13269427/69569532-8656d400-0fc6-11ea-9791-e2b089dd9491.png)

```python
color.grad
```

    tensor([[[ 4.0785e-08,  0.0000e+00,  0.0000e+00],
             [ 2.2414e-07,  0.0000e+00,  0.0000e+00],
             [ 2.6262e-07,  0.0000e+00,  0.0000e+00],
             [-1.4811e-07,  0.0000e+00,  0.0000e+00],
             [-4.1997e-07,  0.0000e+00,  0.0000e+00],
             [-2.6743e-07,  0.0000e+00,  0.0000e+00],
             [-1.1913e-07,  0.0000e+00,  0.0000e+00],
             [-1.3014e-08,  0.0000e+00,  0.0000e+00],
             [ 1.9393e-07,  0.0000e+00,  0.0000e+00],
             [ 5.9817e-08,  0.0000e+00,  0.0000e+00]],
    
            [[ 1.8224e-07,  0.0000e+00,  0.0000e+00],
             [ 1.0772e-06,  0.0000e+00,  0.0000e+00],
             [ 1.2860e-06,  0.0000e+00,  0.0000e+00],
             [-2.5663e-07,  0.0000e+00,  0.0000e+00],
             [-1.5259e-06,  0.0000e+00,  0.0000e+00],
             [-7.8008e-07,  0.0000e+00,  0.0000e+00],
             [-6.9858e-10,  0.0000e+00,  0.0000e+00],
             [-7.7309e-08,  0.0000e+00,  0.0000e+00],
             [ 9.2460e-07,  0.0000e+00,  0.0000e+00],
             [ 2.6062e-07,  0.0000e+00,  0.0000e+00]],
    
            [[ 5.8008e-08,  0.0000e+00,  0.0000e+00],
             [ 7.0951e-07,  0.0000e+00,  0.0000e+00],
             [ 1.1764e-06,  0.0000e+00,  0.0000e+00],
             [ 9.5043e-07,  0.0000e+00,  0.0000e+00],
             [ 2.8123e-07,  0.0000e+00,  0.0000e+00],
             [ 7.1014e-07,  0.0000e+00,  0.0000e+00],
             [ 1.2336e-06,  0.0000e+00,  0.0000e+00],
             [-3.8084e-07,  0.0000e+00,  0.0000e+00],
             [ 5.1185e-07,  0.0000e+00,  0.0000e+00],
             [ 8.3999e-08,  0.0000e+00,  0.0000e+00]],
    
            [[ 4.6083e-10,  0.0000e+00,  0.0000e+00],
             [-1.1402e-07,  0.0000e+00,  0.0000e+00],
             [-1.0701e-07,  0.0000e+00,  0.0000e+00],
             [ 4.8693e-07,  0.0000e+00,  0.0000e+00],
             [ 7.2816e-07,  0.0000e+00,  0.0000e+00],
             [ 6.6526e-07,  0.0000e+00,  0.0000e+00],
             [ 8.1386e-07,  0.0000e+00,  0.0000e+00],
             [-2.3717e-08,  0.0000e+00,  0.0000e+00],
             [-1.1893e-07,  0.0000e+00,  0.0000e+00],
             [ 7.4253e-10,  0.0000e+00,  0.0000e+00]],
    
            [[ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [-4.6290e-08,  0.0000e+00,  0.0000e+00],
             [-2.8072e-07,  0.0000e+00,  0.0000e+00],
             [-2.7842e-07,  0.0000e+00,  0.0000e+00],
             [ 3.7381e-07,  0.0000e+00,  0.0000e+00],
             [ 3.0732e-07,  0.0000e+00,  0.0000e+00],
             [ 1.3061e-07,  0.0000e+00,  0.0000e+00],
             [ 1.9997e-07,  0.0000e+00,  0.0000e+00],
             [-1.8064e-08,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00]],
    
            [[ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [-1.9812e-10,  0.0000e+00,  0.0000e+00],
             [ 5.8225e-08,  0.0000e+00,  0.0000e+00],
             [ 3.0117e-07,  0.0000e+00,  0.0000e+00],
             [ 7.3573e-07,  0.0000e+00,  0.0000e+00],
             [ 5.3496e-07,  0.0000e+00,  0.0000e+00],
             [-2.4259e-07,  0.0000e+00,  0.0000e+00],
             [-1.6055e-07,  0.0000e+00,  0.0000e+00],
             [ 9.1938e-11,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00]],
    
            [[ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [-9.4963e-08,  0.0000e+00,  0.0000e+00],
             [-7.9740e-07,  0.0000e+00,  0.0000e+00],
             [-3.6623e-07,  0.0000e+00,  0.0000e+00],
             [ 2.7589e-07,  0.0000e+00,  0.0000e+00],
             [ 4.4559e-08,  0.0000e+00,  0.0000e+00],
             [-1.0722e-08,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00]],
    
            [[ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [-1.5270e-09,  0.0000e+00,  0.0000e+00],
             [-3.3630e-07,  0.0000e+00,  0.0000e+00],
             [-1.7805e-07,  0.0000e+00,  0.0000e+00],
             [ 2.9738e-07,  0.0000e+00,  0.0000e+00],
             [ 5.0084e-08,  0.0000e+00,  0.0000e+00],
             [ 7.3384e-11,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00]],
    
            [[ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 1.2541e-08,  0.0000e+00,  0.0000e+00],
             [ 4.9522e-07,  0.0000e+00,  0.0000e+00],
             [ 3.1749e-07,  0.0000e+00,  0.0000e+00],
             [ 7.1481e-09,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00]],
    
            [[ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 6.5521e-10,  0.0000e+00,  0.0000e+00],
             [ 9.0145e-08,  0.0000e+00,  0.0000e+00],
             [ 6.6981e-08,  0.0000e+00,  0.0000e+00],
             [ 1.6527e-10,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
             [ 0.0000e+00,  0.0000e+00,  0.0000e+00]]], device='cuda:0')

**Comment by BachiLi** [comment]
Thanks for the detailed report, I'll take a look soon.

**Comment by BachiLi** [comment]
Probably a regression bug after adding the multi-channel texture thing. Stay tuned...

**Comment by BachiLi** [comment]
Should be fixed now. Please pull from the latest commit and try again. Thanks for the report again!
I really should set up integrated testing...if only I have time.

**Comment by nkyriazis** [comment]
Yes! 8d9aaf9e1af5e63700b4b9a3c7facb18fc8e8edd fixed it! Now the example works as expected (very low loss).

Amazing work btw!
